File: proto/src/invite.ts
import type { StorageConfig } from "./storage";

export type WorkspacePayload = {
  id: string;
  name: string;
};

export type InvitePayload = {
  workspace: WorkspacePayload;
  storage: StorageConfig;
};

const INVITE_PREFIX = "bkey-invite:";

export async function generateInvite(
  storage: StorageConfig,
  workspace: WorkspacePayload,
): Promise<string> {
  const payload: InvitePayload = { storage, workspace };
  const b64 = Buffer.from(JSON.stringify(payload)).toString("base64url");
  return `${INVITE_PREFIX}${b64}`;
}

export function parseInvite(link: string): InvitePayload {
  if (!link.startsWith(INVITE_PREFIX)) throw new Error("Invalid invite link");
  try {
    const b64 = link.slice(INVITE_PREFIX.length);
    return JSON.parse(Buffer.from(b64, "base64url").toString("utf-8"));
  } catch {
    throw new Error("Invalid or corrupted invite link");
  }
}

export function applyInvite(link: string): InvitePayload {
  return parseInvite(link);
}

File: envi/envilib/src/invite.rs
use crate::{error::{Error, Result}, storage::StorageConfig};
use base64::{engine::general_purpose::URL_SAFE_NO_PAD as B64URL, Engine};
use serde::{Deserialize, Serialize};

const INVITE_PREFIX: &str = "envi-invite:";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkspacePayload {
    pub id: String,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InvitePayload {
    pub workspace: WorkspacePayload,
    pub storage: StorageConfig,
}

pub fn generate_invite(storage: &StorageConfig, workspace: WorkspacePayload) -> Result<String> {
    let payload = InvitePayload {
        workspace,
        storage: storage.clone(),
    };
    let json = serde_json::to_string(&payload)?;
    let b64 = B64URL.encode(json.as_bytes());
    Ok(format!("{INVITE_PREFIX}{b64}"))
}

pub fn parse_invite(link: &str) -> Result<InvitePayload> {
    let b64 = link
        .strip_prefix(INVITE_PREFIX)
        .ok_or(Error::InvalidInviteLink)?;
    let bytes = B64URL.decode(b64).map_err(|_| Error::InvalidInviteLink)?;
    let json = String::from_utf8(bytes).map_err(|_| Error::InvalidInviteLink)?;
    serde_json::from_str(&json).map_err(|_| Error::InvalidInviteLink)
}
